source/enc_ctx.c

```diff
@@ -64,29 +64,29 @@ int aws_cryptosdk_enc_ctx_size(size_t *size, const struct aws_hash_table *enc_ct
 
 int aws_cryptosdk_enc_ctx_serialize(
     struct aws_allocator *alloc, struct aws_byte_buf *output, const struct aws_hash_table *enc_ctx) {
-    size_t length;
-    if (aws_cryptosdk_enc_ctx_size(&length, enc_ctx)) {
-        return AWS_OP_ERR;
-    }
+    AWS_PRECONDITION(AWS_OBJECT_PTR_IS_READABLE(alloc));
+    AWS_PRECONDITION(aws_byte_buf_is_valid(output));
+    AWS_PRECONDITION(aws_hash_table_is_valid(enc_ctx));
+
+    size_t num_elems = aws_hash_table_get_entry_count(enc_ctx);
+    if (num_elems > UINT16_MAX) return aws_raise_error(AWS_CRYPTOSDK_ERR_LIMIT_EXCEEDED);
 
-    if (output->capacity < length) {
+    if (!aws_byte_buf_write_be16(output, (uint16_t)num_elems)) {
+        aws_byte_buf_clean_up(output);
```

Comment:
What is the value of invoking clean up just before raising error?  How does this help the proof of memory safety or functional correctness?

Reply:
This is on the actual code.  We need to avoid memory leaks.